### PR TITLE
Use coverage from data collector

### PR DESCRIPTION
**Pull request by tonyhallett**
Given that runsettings is now being respected so too should coverlet data collection.

This is one way in which https://github.com/FortuneN/FineCodeCoverage/issues/30#issuecomment-751455190 can be resolved.

It also has the benefit of not running tests twice.  Why wait longer than is necessary for coverage ? More importantly is the effect that it has for ui tests.

Note that I have added some reflection helpers which can be extended for TestExecutionFinished.

### CODE REVIEW

**Comment by tonyhallett** [comment]
@FortuneN what do you think ?
I may have missed some error checking code.
If you like what you see the reflection wrappers can be extended for TestExecutionFinished.
If you would rather this to be an opt-in experimental feature then I can add to options.
Perhaps it could be tested against the project from @Yaevh https://github.com/FortuneN/FineCodeCoverage/issues/30#issuecomment-751415343

**Comment by tonyhallett** [comment]
Just remembered that there is a small piece of code to be implemented in it. 
https://github.com/tonyhallett/FineCodeCoverage/blob/a304ba02c66b3a166ee35237c63bc7a0d19fa4cd/FineCodeCoverage/Core/Coverlet/RunSettingsResultsDirectoryForEnabledCollectorParser.cs#L48

The project I was running against had the following runsettings with a relative path.

```
<RunSettings>
  <RunConfiguration>
    <ResultsDirectory>.\MyTestResults</ResultsDirectory>
    <TestAdaptersPaths>C:\Users\tonyh\.nuget\packages\coverlet.collector\3.0.0-preview.16.g20b94e8b54\build\netstandard1.0</TestAdaptersPaths>
  </RunConfiguration>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="XPlat code coverage"/>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```

I will add a commit for that tomorrow.

**Comment by FortuneN** [comment]
@tonyhallett 
I love this. It is some super cool engineering.

I am however reluctant to merge this in it's current form.

Remember, i have plans to uproot the the 'Core' folder into a webserver.
This feature, in its current form becomes a hinderance to that goal.

Don't get me wrong. I want us to keep the feature but we are going to have to make a few architectural changes.

What i'm proposing is:

1. Create a new project for the web-server [**FineCodeCoverageMicroServer**]
Move the 'Core' stuff into the web-server.
The web-server will expose our current core functionality over HTTP.
The web-server should be distributable as a single file executable.
This will allow us to be multi-platform and integrate into other IDEs the need not be written in any CLR language (HTTP is the neutral interface).
(I'm looking a VSCode as the next IDE)
**_.NetCore_**

2. The current FineCodeCoverage project will embed the web-server and make calls there for our current needs. (Calculating coverage and obtaining results and the html report)

3. The feature detailed in this PR will be in the current FineCodeCoverage project.
I'm thinking of it as a platform-native optimization as i'm sure we will discover other platform-specific optimizations when we start working on extensions for other IDEs.

**In short, I'd like us to have our cake and eat it too.**

I will create a separate branch to start working on this.

**Comment by tonyhallett** [comment]
@FortuneN 

> The feature detailed in this PR will be in the current FineCodeCoverage project.
I'm thinking of it as a platform-native optimization as i'm sure we will discover other platform-specific optimizations when we start working on extensions for other IDEs.

Yes, if you go that route then it would be platform-native.

Unfortunately, although this worked perfectly against one of my solutions, it failed to work with @Yaevh https://github.com/Yaevh/fcc-net50-bug-repro-2.

Additionally, the collector version 1.3 does not surface the coverage file into the TestResults folder.  The nightly version of the collector does.

I will look at it a bit further as I strongly feel ( as I am sure you know ) that using the global tool is a bad idea and duplicating tests and waiting for coverage is too.
